### ISSUE TITLE
Add an example that assembles a training run from the core pieces

### ISSUE THREAD

**Issue by amazloumi** [KempnerInstitute/KempnerForge#186]
Follow-up from the review of #178.

`run_training` builds a job's phases in a required order, and the order is not
self-evident from the pieces. Add a standalone example under `examples/` that
assembles a run directly from the building blocks instead of calling
`run_training`, and writes the ordering constraints down.

Own folder, not inside an existing experiment example: the constraints are
model-agnostic, and every example has to stay independently deletable.

## Why it matters

Violating the order does not crash. The clearest case is dataloader state.
`restore_checkpoint` runs before the dataloader exists, so `load()` stashes the
loader state and a separate call applies it once the loader is built:

```python
data = build_data_pipeline(config, runtime)
if data.dataloader is not None:
    ckpt_mgr.apply_dataloader_state(data.dataloader)
```

Omit that second call and the model, optimizer and scheduler still resume
correctly, the run still logs that it resumed from step N, and it still exits 0
— but the dataloader restarts from the beginning and the run re-trains on
batches it has already seen. The only symptom is a log line that is missing.

## Scope

- A minimal text-model run wired up step by step.
- The ordering constraints stated where they are needed, including: the
  checkpoint manager registers model/optimizer/scheduler and so needs them
  first; phase scheduling restored by `load()` determines loader identity, so
  the loader is built after restore; `build_phase_state` needs the restored
  step; teardown collectives belong on the success path only.
- A test that omits `apply_dataloader_state`, resumes, and asserts the
  dataloader position diverges — so the constraint is enforced, not just
  described.

## Open question

If this is easy enough to get wrong, the stronger fix may be in core: detect
stashed dataloader state that was never applied and fail loudly. That would make
the example documentation rather than the only defense. Worth deciding before
the example is written.
